lib/pbench/server/cache_manager.py

```diff
@@ -226,6 +295,103 @@ def create(cls, tarball: Path, controller: "Controller") -> "Tarball":
+            }
+        }
+        dir_queue = deque([(dir_path, cmap)])
+        while dir_queue:
+            dir_path, parent_map = dir_queue.popleft()
+            tar_n = dir_path.name
+
+            curr = {}
+            for l_path in dir_path.glob("*"):
+                tar_info = CacheObject(dir_path=root_dir_path, path=l_path)
+                curr[l_path.name] = {"details": tar_info}
+                if l_path.is_dir():
+                    dir_queue.append((l_path, curr))
+            parent_map[tar_n]["children"] = curr
+
+        self.cachemap = cmap
+
+    @staticmethod
+    def traverse_cmap(path: Path, cachemap: dict) -> dict[str, dict]:
+        """Sequentially traverses the cachemap to find the leaf of a
+        relative path reference
+
+        Args:
+            path: relative path of the sub-directory/file
+            cachemap: dictionary mapping of the root Dicrectory
+
+        Raises:
+            BadDirpath if the directory/file path is not valid
+
+        Returns:
+            Dictionary with directory/file details or children if present
+        """
+        file_list = path.parts[:-1]
+        f_entries = cachemap
+
+        try:
+            for file_l in file_list:
+                info = f_entries[file_l]
+                if info["details"].type == "DIRECTORY":
+                    f_entries = info["children"]
+                else:
+                    raise BadDirpath(
+                        f"Found a file {file_l!r} where a directory was expected in path {path}"
+                    )
+
+            return f_entries[path.name]
+        except KeyError as exc:
+            raise BadDirpath(f"File/directory {exc} in path {path} not found in cache.")
+
+    def get_info(self, path: Path) -> dict[str, dict]:
+        """Returns the details of the given file/directory in dict format
+
+        Args:
+            path: path of the file/sub-directory
+
+        Raises:
+            BadDirpath on bad directory path
+
+        Returns:
+            Dictionary with Details of the file/directory
+        """
+        if str(path).startswith("/"):
+            raise BadDirpath(
+                f"The directory path {str(path)!r} is an absolute path,"
+                " we expect relative path to the root directory."
+            )
+
+        c_map = self.traverse_cmap(path, self.cachemap)
+        children = c_map["children"] if "children" in c_map else {}
+        fd_info = c_map["details"].__dict__.copy()
+
+        if fd_info["type"] == "DIRECTORY":
+            fd_info["directories"] = []
+            fd_info["files"] = []
+
+            for key, value in children.items():
+                if value["details"].type == "DIRECTORY":
+                    fd_info["directories"].append(key)
+                elif value["details"].type == "FILE":
+                    fd_info["files"].append(key)
+
+            fd_info["directories"].sort()
+            fd_info["files"].sort()
+
+        return fd_info
```

Comment:
JSON is really flexible, and having APIs return JSON makes sense; I've also used it for internal information (e.g., the `ApiContext`), where the information required by various APIs varies. But where we have fixed data that's used internally, objects with named attributes are easier to use and less error-prone. I'm happy to see the cache entry dataclass, but I thought we were going a step further, to build the tree out of classes and to return some sort of `FileInfo` instance here. Again, this avoids typing literal strings everywhere.